# Module 60.1: Yield-Based Bond Convexity and Portfolio Properties
<hr>

In [25]:
# Import libraries
import math
import pandas as pd
import numpy as np
import numpy_financial as npf
# import matplotlib.pyplot as plt
# import seaborn as sns
# from sklearn.model_selection import train_test_split
# from pandas.core.common import random_state
# from sklearn.linear_model import LinearRegression

### LOS 60.a: Calculate and interpret convexity and describe the convexity adjustment.

In our reading on Yield-Based Bond Duration Measures and Properties, we showed that modified duration is a linear approximation of the price-yield relationship. Because the true relationship is convex, duration-based estimates of a bond's full price will be increasingly different from actual prices as we increase the changes in yield. This is illustrated in  Price-Yield Curve for an Option-Free, 8%, 20-Year **Semiannual-Pay Bond**. Duration-based price estimates for a decrease and for an increase in YTM are shown as Est.– and Est.+.

*Price-Yield Curve for an Option-Free, 8%, 20-Year Semiannual-Pay Bond*

<img src="https://github.com/PachaTech/CFA-Level-1/blob/main/Fixed%20Income/Module%2060/pics/par.jpeg?raw=true"></img>

We can improve our estimates of the price impact of a change in yield by introducing a second term. Convexity is a measure of the curvature of the price-yield relation. The more curved it is, the greater the convexity adjustment.

One way to calculate convexity is by considering each of a bond's cash flows separately. The convexity of a single cash flow at period t is given by the following:

$\text{convexity of cash flow at period } t = t×t+1(1+r)^2$

where $r$ is the periodic yield of the cash flow $\large\frac{\text{YTM}}{\text{periodicity}}$

The convexity of a coupon-paying bond can then be calculated as the weighted average convexity of its individual cash flows, using the present value of cash flows as the weights (the same weighting we use to calculate Macaulay duration).

For a 5-year, 11% annual coupon bond priced at 86.59 (a 15% yield to maturity), the convexity of the coupon at 

`Time 1 is (1 × 2) / 1.152 = 1.512`, 
\
and the convexity of the coupon at 
\
`Time 2 is (2 × 3) / 1.152 = 4.537`. 

The following table completes the calculations for all five of the bond's cash flows:

<img src="https://github.com/PachaTech/CFA-Level-1/blob/main/Fixed%20Income/Module%2060/pics/1.png?raw=true"></img>


Using the weights given in the table, we can calculate the convexity for this bond as follows:

$$
0.1105(1.512) + 0.0961(4.537) + 0.0835(9.074) + 0.0726(15.123) + 0.6373(22.684) = 16.915
$$

For bonds with non-annual coupons, convexity needs to be divided by the number of periods per year *squared to annualize* the measure. For a semiannual coupon bond, the final convexity figure would be annualized by dividing by $2^2 = 4$.

In a similar way to how we approximated modified duration, we can also determine a bond's approximate convexity using the following formula:

$$
\text{approximate convexity} = \frac{V_−+V_+−2V_0}{(\Delta \text{YTM})^2 V_0}
$$

where:

$V_{–}$ = price of the bond if YTM is decreased by $\Delta$ YTM

$V_{+}$ = price of the bond if the YTM is increased by $\Delta$ YTM

$V_{0}$ = current price of the bond

**Example: Calculating approximate convexity**

Consider our 5-year, 11% annual coupon bond priced at 86.59138 to yield 15% to maturity. If its YTM increases by 50 basis points, its price will decrease to 85.09217. If its YTM decreases by 50 basis points, its price will increase to 88.12721. Calculate the approximate convexity of the bond.

In [64]:
# Example: Calculating approximate convexity

print(float(88.12721 + 85.09217) - (2*86.59138))
print(float(86.59138 * (0.005**2)))
v = float(0.03661999999999921/0.0021647845)

print("Bond approximate convexity is: ", round(v,3))

0.03661999999999921
0.0021647845
Bond approximate convexity is:  16.916


**Answer:**

The approximate convexity is: 
$\large\frac{88.12721+85.09217–(2×86.59138)}{86.59138×0.005^2}\small=16.916$

This result is nearly the same as the convexity measure we calculated by considering each cash flow separately.

A bond's convexity is increased or decreased by the same bond characteristics that affect duration. A longer maturity, a lower coupon rate, or a lower YTM will all increase convexity, and vice versa. For two bonds with equal duration, the one with cash flows that are more dispersed over time will have more convexity.

### LOS 60.b: Calculate the percentage price change of a bond for a specified change in yield, given the bond's duration and convexity.

By taking account of both a bond's duration (first-order effects) and convexity (second-order effects), we can improve an estimate of the effects of a change in yield on a bond's value, especially for larger changes in yield:

$$
\text{percent change in full bond price} = \text{–annual modified duration} (\Delta \text{YTM}) + \frac{1}{2} \text{annual convexity} (\Delta \text{YTM})^2
$$

<hr>

**Example: Estimating price changes with duration and convexity**

Consider our 5-year, 11% annual coupon bond priced at 86.59138 to yield 15% to maturity. We have calculated the modified duration to be 3.50 and the convexity of the bond to be 16.9. Estimate the new price of the bond if its yield decreases by 50 basis points.


In [44]:
# Estimating price changes with duration and convexity

# The duration effect
dur_effect = 3.50 * 0.005
print(dur_effect)

# The convexity effect is
conv_effect = 0.5 * 16.9 * (0.005**2) *100
print(round(conv_effect,4))

# The expected change in bond price is 
exp_change =  (.0175 + .000211) *100
print(round(exp_change,4))

# The new price of the bond is estimated to be
bond_newp = 86.59138 * 1.017711
print(round(bond_newp,4))



0.0175
0.0211
1.7711
88.125



**Answer:**

The duration effect is –3.50 × –0.005 = 1.75%.

The convexity effect is 12×16.9×(−0.005)2=0.000211=0.0211%.

The expected change in bond price is (1.75% + 0.0211%) = 1.7711%.

The new price of the bond is estimated to be 86.59138 × 1.017711 = 88.125.

Analogous to money duration (`MoneyDur`), the money convexity (`MoneyCon`) of a bond position is expressed in currency units:

$\text{money convexity} = \text{annual convexity} × \text{full price of bond position}$

We can use money duration and money convexity to estimate the change in price of a bond as follows:

$\text{change in full price of bond} = \text{–(MoneyDur } x  \Delta \text{(YTM)} + \frac{1}{2} x \text{ MoneyCon } x \Delta\text{YTM)}^2)$


<hr>

**Example: Estimating price changes with duration and convexity**

For the bond in our previous examples, calculate the money duration and money convexity of a $10 million par position in the bond and estimate the new price of the bond for a 50 basis point decrease in yield. Recall that the modified duration of the bond is 3.50 and its convexity is 16.9.


In [76]:
# The market value of the position is 
print("The market value of the position is   = $",0.8659138 * 10000000)

# The money duration of the position is 
print("The money duration of the position is = $",3.50 * 8659138)

# The money convexity of the bond is 
print("The money convexity of the bond is    = $",16.9 * 8659138)

# The duration effect is 
print("The duration effect is                = $",round(-(30306983 * -0.005),2))

# The convexity effect is  
print("The convexity effect is               = $",round(0.5*146339432 * (0.005**2),2)) 

# The expected change in bond price is 
print("The expected change in bond price is  = $",151534.92 + 1829.25)

# The new value of the bond is estimated to be 
print("The new estimated value of the bond   = $",8659138 + 153364.17)

The market value of the position is   = $ 8659138.0
The money duration of the position is = $ 30306983.0
The money convexity of the bond is    = $ 146339432.2
The duration effect is                = $ 151534.92
The convexity effect is               = $ 1829.24
The expected change in bond price is  = $ 153364.17
The new estimated value of the bond   = $ 8812502.17


**Answer:**

The market value of the position is \\$0.8659138 × 10,000,000 = \\$8,659,138.

The money duration of the position is \\$3.50 × \\$8,659,138 = \\$30,306,983.

The money convexity of the bond is \\$16.9 × \\$8,659,138 = \\$146,339,432$.

The duration effect is –(\\$30,306,983 × –0.005) = \\$151,534.92.

The convexity effect is $\frac{1}{2} × \$146,339,432 × (–0.005)^2 = \$1,829.25$

The expected change in bond price is (\\$151,534.92 + \\$1,829.25) = \\$153,364.17.

The new value of the bond is estimated to be \\$8,659,138 + \\$153,364.17 = \\$8,812,502. This is consistent with the estimate for new price of 88.125 in the previous example. Both methods are doing the same thing in a slightly different way.

The convexity adjustment to the price change is the same for either an increase or a decrease in yield. As illustrated in  **Duration-Based Price Estimates vs. Actual Bond Prices**, the duration-only based estimate of the increase in price resulting from a decrease in yield is too low for a bond with positive convexity, and it is improved by a positive adjustment for convexity. The duration-only based estimate of the decrease in price resulting from an increase in yield is larger than the actual decrease, so it is also improved by a positive adjustment for convexity.


*Duration-Based Price Estimates vs. Actual Bond Prices*

<center>
<img src="https://github.com/PachaTech/CFA-Level-1/blob/main/Fixed%20Income/Module%2060/pics/2.jpeg?raw=true"></img>
</center>

### LOS 60.c: Calculate portfolio duration and convexity and explain the limitations of these measures.

There are two approaches to estimating the duration and convexity of a portfolio. The first is to base a single duration and convexity calculation on the portfolio's aggregate cash flows across all bonds. The second approach is to calculate the duration and convexity of each bond in the portfolio, then take a weighted average of these based on the weight of the bond in the portfolio.

The first approach is theoretically correct; however, the second approach is typically used in practice because it is often easier to apply. The second approach can be formulated as follows:

$$
\text{portfolio duration} = W_1 D_1 + W_2 D_2 + \ldots + W_N D_N
$$

where:

$W_i$ = full price of bond $i$ divided by the total value of the portfolio

$D_i$ = duration of bond $i$

$N$ = number of bonds in the portfolio

One limitation of this approach is that it assumes the YTM of every bond (of any maturity) in the portfolio must change by the same amount. Only under this assumption of a **parallel shift** in the yield curve will portfolio duration and convexity calculated with this approach produce the percentage change in portfolio value per 1/% change in YTM.

Changes in the yield curve are rarely a simple parallel shift. Changes in shape, such as a steepening or twist, are common.

## Reading 60: Key Concepts:

<hr>

### LOS 60.a

Convexity refers to the curvature of a bond's price-yield relationship.

The convexity of a single cash flow at period t is given by the following:

convexity of cash flow at period $t = \Large\frac{t×(t+1)}{(1+r)^2}$

where:

$t$ = period at which the cash flow occurs

$r$ = periodic yield of the bond (YTM/periodicity)

The convexity of a coupon-paying bond is the weighted average convexity of its cash flows.

To annualize convexity for non-annual coupons, divide by periodicity squared.

Convexity can be approximated using the following formula:

approximate convexity=V−+V+−2V0(ΔYTM)2V0

where:

$V_–$ = price of the bond if YTM is decreased by ΔYTM

$V_+$ = price of the bond if the YTM is increased by ΔYTM

$V_0$ = current price of the bond

### LOS 60.b

Given values for approximate annual modified duration and approximate annual convexity, the percentage change in the full price of a bond can be estimated as follows:

$$
\%\Delta \text{full bond price  } = \text{-annual modified duration (VYTM)} + \frac{1}{2} \text{  annual convexity (VYTM)}^2
$$

Money convexity is stated in currency units and is sometimes expressed per 100 of bond value:

money convexity = annual convexity × full price of bond position

Using money duration and money convexity to directly estimate the change in price of a bond, this is the equation:

$$
\text{change in full price of bond} = \big(\text{-MoneyDur} × \Delta \text{YTM}\big)+ \Big(\frac{1}{2} × \text{MoneyCon} × \Delta \text{YTM}^2\Big)
$$

### LOS 60.c

There are two methods for calculating portfolio duration and convexity:

1. Calculate a single duration and convexity measure based on the aggregate cash flows of the bond portfolio.
2. Calculate the weighted average of durations of bonds in the portfolio. This method is used most often in practice, but it assumes a parallel shift of the yield curve.

## Module 60.1: Yield-Based Bond Convexity and Portfolio Properties Quiz

<hr>

#### QUESTION 1 OF 5
**Note** -- this is also another question using a Macauly table (Matrix) unlikely used on the exam.


The annualized convexity of a 2-year, annual-pay, 10% coupon bond trading at par is closest to:


* **ANSWER:** \
4.66.

* **EXPLANATION** \
Because the bond is trading at par, its yield is equal to its coupon of 10%. The convexity of the first cash flow at Time 1 is (1 × 2) / (1.10)2 = 1.653. The convexity of the second cash flow is (2 × 3) / (1.10)2 = 4.959. The weights for each cash flow time are calculated as follows:
+ The convexity of the bond is 0.09091(1.653) + 0.90909(4.959) = 4.66.



In [60]:
# Question 1 of 5 - Annualized convexity question

t1 = (1*2)/(1.10**2)
t2 = (2*3)/(1.10**2)

Q1convexity = (0.09091 * t1) + (0.90909 * t2)

print(round(t1,3))
print(round(t2,3))
print("")
print("The Annualized convexity of the bond is =",round(Q1convexity,3))

1.653
4.959

The Annualized convexity of the bond is = 4.658


#### QUESTION 2 OF 5

A bond is trading at a price of 104.4518. If its yield increases by 10 basis points, its price will decrease to 103.9954. If its yield decreases by 10 basis points, its prices will increase to 104.9108. The approximate convexity of this bond is closest to:


* **ANSWER:** \
24.89.

* **EXPLANATION** \
The approximate convexity of the bond is calculated as follows:
$$
\text{approximate convexity} = \frac{V_−+V_+−2V_0}{(\Delta \text{YTM})^2 V_0}
$$

where:

$V_{–}$ = price of the bond if YTM is decreased by $\Delta$ YTM

$V_{+}$ = price of the bond if the YTM is increased by $\Delta$ YTM

$V_{0}$ = current price of the bond

In this case, 

$$
\text{approximate convexity} = \frac{104.9108+103.9954 - 2\times(104.4518)}{0.001^2 \times 104.4518} = 24.98 
$$

In [48]:
# QUESTION 2 OF 5 - approximate convexity question

a = 104.9108+103.9954-(2*(104.4518))
b = (0.001**2)*104.4518
solution = a/b

print(104.9108+103.9954-(2*(104.4518)))
print((0.001**2)*104.4518)
print("")
print("the approximate convexity is = ", round(solution,2))

0.002600000000001046
0.0001044518

the approximate convexity is =  24.89


#### QUESTION 3 OF 5
Note: This uses a Macauly matrix and it is unlikely on the level 1 exam.

The annualized convexity of a 2-year, semiannual-pay, 10% coupon bond trading at par is closest to:


* **ANSWER:** \
4.12.

* **EXPLANATION** \
Because the bond is trading at par, its yield is equal to its coupon of 10\%. Note that this is a semiannual coupon bond; hence, it has coupons of //$5 every six months and has a periodic six-month return of 5\%. The weights for each cash flow time are calculated as follows:

1. The convexity of the first cash flow at Time 1 is (1 × 2) / (1.05)2 = 1.8141.

2. The convexity of the second cash flow is (2 × 3) / (1.05)2 = 5.4422.

3. The convexity of the third cash flow is (3 × 4) / (1.05)2 = 10.8844.

3. The convexity of the fourth cash flow is (4 × 5) / (1.05)2 = 18.1406.

* The convexity of the bond is 0.0476(1.8141) + 0.0454(5.4422) + 0.0432(10.8844) + 0.8638(18.1406) = 16.47.

* The annualized convexity is calculated by dividing convexity by the periodicity of the bond (2) squared.

* Hence, annualized convexity = (16.47 / 2)^2 = 4.12.

(Module 60.1, LOS 60.a)

In [33]:
# Question 3 of 5

convexity = 0.0476*(1.8141) + 0.0454*(5.4422) + 0.0432*(10.8844) + 0.8638*(18.1406)
print(round(convexity,2))
annualconvexity = (convexity / 2**2)
print(round(annualconvexity,2))

16.47
4.12


#### QUESTION 4 OF 5

A bond has a convexity of 114.6. The convexity effect, if the yield decreases by <u>110 basis points</u>, is closest to:

* **ANSWER:** \
+0.693\%.

* **EXPLANATION** \
$$
\text{The convexity effect} = 12 × \text{convexity} × (\Delta \text{YTM})^2 \\
= (0.5)(114.6)(-0.011)^2= 
0.00693= 0.693%.
$$
\
(Module 60.1, LOS 60.b)

In [14]:
# QUESTION 4 OF 5
# A bond has a convexity of 114.6. 
# The convexity effect, if the yield decreases by 110 basis points, 
# is closest to:

# USE CONVEXITY EFFECT FORMULA

q4 = (0.5)*(114.6)*(-0.011**2)*-100 
print(round(q4,3))

0.693


#### QUESTION 5 OF 5
Portfolio duration based on weighted average durations of constituents of the portfolio assumes:


* **ANSWER:** \
yields change uniformly across all maturities.

* **EXPLANATION** \
Portfolio duration is limited as a measure of interest rate risk because it assumes parallel shifts in the yield curve; that is, the discount rate at each maturity changes by the same amount. 
\
(Module 60.1, LOS 60.c)


## Reading 60 QBank 1

<hr>


#### QUESTION 1 OF 6
The annual convexity of a bond is calculated as 12.35. If the full price of the bond position is \\$1.5 million and the bond matures in three years, the money convexity is *closest* to:


* **ANSWER:** \
\\$18,525,000

* **EXPLANATION** \
The money convexity of a bond is equal to its annual convexity times the full price of the bond position. With an annual convexity of 12.35 and the full price of the bond position of \\$1,500,000, the money convexity is equal to 12.35 × \\$1,500,000 = \\$18,525,000. The maturity of the bond does not impact the calculation.
$12.34 \times 1,500,000 = 18,525,000$
(Module 60.1, LOS 60.b)



In [72]:
print(12.34 * 1500000)

18510000.0


<hr>

#### QUESTION 2 OF 6
A fixed-income portfolio manager is estimating portfolio duration based on the weighted average of the durations of each bond in the portfolio. The manager should calculate duration using:


* **ANSWER:** \
parallel shifts of the benchmark yield curve.

* **EXPLANATION** \
Portfolio duration as a weighted average of the individual bonds' durations is calculated assuming parallel shifts in the yield curve. Cash flow yield is used to calculate duration based on the weighted average time until a bond portfolio's cash flows are scheduled to be received.\
(Module 60.1, LOS 60.c).

#### QUESTION 3 OF 6
An annual-pay bond is priced at 101.50. If its yield to maturity decreases 100 basis points, its price will increase to 105.90. If its yield to maturity increases 100 basis points, its price will decrease to 97.30. The bond's approximate modified convexity is closest to:


* **ANSWER:** \
parallel shifts of the benchmark yield curve.

* **EXPLANATION** \
The approximate convexity of the bond is calculated as follows:
$$
\text{approximate convexity} = \frac{V_−+V_+−2V_0}{(\Delta \text{YTM})^2 V_0}
$$

where:

$V_{–}$ = price of the bond if YTM is decreased by $\Delta$ YTM

$V_{+}$ = price of the bond if the YTM is increased by $\Delta$ YTM

$V_{0}$ = current price of the bond

In this case, 

$$
\text{modified convexity} = \frac{105.90+97.30 - 2\times(101.50)}{0.01^2 \times 101.5} = 19.70
$$



In [73]:
# QUESTION 2 OF 5 - approximate convexity question

a2 = 105.90+97.30-(2*(101.5))
b2 = (0.01**2)*101.5
solution2 = a2/b2

print(a2)
print(b2)
print("")
print("the approximate convexity is = ", round(solution2,2))

0.19999999999998863
0.010150000000000001

the approximate convexity is =  19.7


<hr>

#### QUESTION 3 OF 6
A bond is priced at 95.80. Using a pricing model, an analyst estimates that a 25 bp parallel upward shift in the yield curve would decrease the bond's price to 94.75, while a 25 bp parallel downward shift in the yield curve would increase its price to 96.75. The bond's effective convexity is closest to:

* **ANSWER:** \
-167.

* **EXPLANATION** \
The approximate convexity of the bond is calculated as follows:
$$
\text{approximate convexity} = \frac{V_−+V_+−2V_0}{(\Delta \text{YTM})^2 V_0}
$$

where:

$V_{–}$ = price of the bond if YTM is decreased by $\Delta$ YTM

$V_{+}$ = price of the bond if the YTM is increased by $\Delta$ YTM

$V_{0}$ = current price of the bond

In this case, 

$$
\text{modified convexity} = \frac{96.75 + 94.75 - 2\times(95.80)}{0.0025^2 \times 95.80} = 167.01
$$

In [74]:
# QUESTION 3 OF 5 - approximate convexity question

a3 = 96.75+94.75-(2*(95.80))
b3 = (0.0025**2)*95.80
solution3 = a3/b3

print(a3)
print(b3)
print("")
print("the approximate convexity is = ", round(solution3,2))

-0.09999999999999432
0.00059875

the approximate convexity is =  -167.01


<hr>

#### QUESTION 5 OF 6
Negative convexity is most likely to be observed in:

* **ANSWER:** \
callable bonds.

* **EXPLANATION** \
All noncallable bonds exhibit the trait of being positively convex. Callable bonds have negative convexity because once the yield falls below a certain point prices will rise at a decreasing rate, thus giving the price-yield relationship a negative convex shape.\
(Module 60.1, LOS 60.a).

<hr>

#### QUESTION 6 OF 6
Rather than using the annual modified duration and convexity numbers, a financial analyst plans to use money duration and money convexity to estimate the new price of a bond after a change in yield. Relative to the estimated price using the annual numbers, the "money" duration and convexity will produce an estimated bond price that is:

* **ANSWER:** \
equivalent.

* **EXPLANATION** \
Although the calculation incorporates different components, the estimated price change and new bond price will be equivalent under the approach that uses annual duration and convexity versus the approach that uses money duration and convexity. Any differences would be due to rounding.
\
(Module 60.1, LOS 60.b)